lib/stage.ts

```diff
@@ -0,0 +1,231 @@
+					sourceStage: stageIdx,
+					...copy,
+				}),
+			);
+
+			this.actionGroups.push(actionGroup);
+			this.ungroupedCommands = [];
+		}
+
+		this.lastStepWasCopy = true;
+
+		this.dependentOnStages = _.uniq(this.dependentOnStages.concat(stageIdx));
+	}
+
+	public addCommandStep(command: string) {
+		this.ungroupedCommands.push(command);
+		this.lastStepWasCopy = false;
+	}
+
+	public addWorkdirStep(workdir: string) {
+		// We need to create a new group, saving any ungrouped commands into
+		// the latest group
+		const actionGroup = _.last(this.actionGroups) as ActionGroup;
+		actionGroup.commands = actionGroup.commands.concat(this.ungroupedCommands);
+		this.actionGroups.push({
+			copies: [],
+			dependentOnStage: false,
+			commands: [],
+			workdir,
+		});
+		this.lastWorkdir = workdir;
+		this.ungroupedCommands = [];
+		this.lastStepWasCopy = false;
+	}
+
+	public finalize() {
+		const actionGroup = _.last(this.actionGroups) as ActionGroup;
+		actionGroup.commands = actionGroup.commands.concat(this.ungroupedCommands);
+
+		// Also filter any action groups which do not have commands or copies,
+		// as these are pointless
+		this.actionGroups = _.reject(
+			this.actionGroups,
+			ag => ag.commands.length === 0 && ag.copies.length === 0,
+		);
+	}
+
+	public getActionGroupsForChangedFiles(files: string[]): ActionGroup[] {
+		// Go through the action groups in order, checking if the file should
+		// trigger the group, and returning every action group that follows
+		for (const [idx, actionGroup] of this.actionGroups.entries()) {
+			if (actionGroup.dependentOnStage) {
+				// We're not interested in stage copies
+				continue;
+			}
+			const matches = fileMatchesForActionGroup(files, actionGroup);
+			if (matches.length > 0) {
+				// return this + everything after it
+				return this.actionGroups.slice(idx);
+			}
+		}
+		return [];
+	}
+
+	public getActionGroupsForChangedStage(stageIdx: number): ActionGroup[] {
+		for (const [idx, actionGroup] of this.actionGroups.entries()) {
+			if (!actionGroup.dependentOnStage) {
+				// We're not interested in non-stage copies
+				continue;
+			}
+			if (actionGroup.stageDependency === stageIdx) {
+				return this.actionGroups.slice(idx);
+			}
+		}
+
+		return [];
+	}
+
+	private canAddCopyToGroup(dependentOnStage: false): boolean;
+	private canAddCopyToGroup(dependentOnStage: true, stageIdx: number): boolean;
+	private canAddCopyToGroup(
+		dependentOnStage: boolean,
+		stageIdx?: number,
+	): boolean {
+		if (this.lastStepWasCopy) {
+			const actionGroup = _.last(this.actionGroups) as ActionGroup;
+			// the last step's last addition was a copy. We can add it to the
+			// same group iff the stage dependency matches (or lack thereof)
+			if (dependentOnStage) {
+				return (
+					actionGroup.dependentOnStage &&
+					actionGroup.stageDependency === stageIdx
+				);
+			}
+			return !actionGroup.dependentOnStage;
+		}
+		return false;
+	}
+
+	private copyArgsToCopies(
+		args: string[],
```

Comment:
This would be better typed as `[ string, string, ...string[] ]` to type that a minimum of two strings must be passed

Reply:
Didn't think of that, nice!